src/compiler/propagation/type_propagator.cc

```diff
@@ -1067,17 +1067,27 @@ static TypeScope* process(TypeScope* scope, uint8* bcp, std::vector<Worklist*>&
   OPCODE_END();
 
   OPCODE_BEGIN(BRANCH_IF_TRUE);
+    TypeSet top = stack->local(0);
+    bool can_be_falsy = top.contains_null(program) || top.contains_false(program);
```

Comment:
I think you could do the same for truthy.  If the only types present are in the set {null, false} then it can't be truthy.